gengo/src/lib.rs

```diff
@@ -39,127 +42,243 @@ const GLOB_MATCH_OPTIONS: MatchOptions = MatchOptions {
                 }
-                _ => continue,
-            }
-        }
+                let Ok(path) =
+                    gix::path::try_from_bstr(entry.index_entry.path_in(&results.path_storage))
+                else {
+                    return Ok(());
+                };
+                self.analyze_blob(path, repo, state, entry)
+            },
+            || Some(std::time::Duration::from_micros(5)),
+            std::convert::identity,
+        )?;
         Ok(())
     }
 
-    fn analyze_blob<P: AsRef<Path>>(
+    fn analyze_blob(
         &self,
-        filepath: P,
-        blob: &Blob,
-        results: &mut IndexMap<PathBuf, Entry>,
+        filepath: impl AsRef<Path>,
+        repo: &gix::Repository,
+        state: &mut GitState,
+        result: &mut BlobEntry,
     ) -> Result<()> {
-        let contents = blob.content();
+        let filepath = filepath.as_ref();
+        let blob = repo.find_object(result.index_entry.id)?;
+        let contents = blob.data.as_slice();
+        state
+            .attr_stack
+            .at_path(filepath, Some(false), |id, buf| {
+                repo.objects.find_blob(id, buf)
+            })?
+            .matching_attributes(&mut state.attr_matches);
 
-        let lang_override = self
-            .get_str_attr(&filepath, "gengo-language")?
-            .map(|s| s.replace('-', " "))
+        let mut attrs = [None, None, None, None, None];
+        state
+            .attr_matches
+            .iter_selected()
+            .zip(attrs.iter_mut())
+            .for_each(|(info, slot)| {
+                *slot =
+                    (info.assignment.state != gix::attrs::StateRef::Unspecified).then_some(info);
+            });
+
+        let lang_override = attrs[0]
+            .as_ref()
+            .and_then(|info| match info.assignment.state {
+                StateRef::Value(v) => v.as_bstr().to_str().ok().map(|s| s.replace('-', " ")),
+                _ => None,
+            })
             .and_then(|s| self.analyzers.get(&s));
 
         let language =
-            lang_override.or_else(|| self.analyzers.pick(&filepath, contents, self.read_limit));
+            lang_override.or_else(|| self.analyzers.pick(filepath, contents, self.read_limit));
 
         let language = if let Some(language) = language {
-            language.clone()
+            language
         } else {
             return Ok(());
         };
 
-        let size = contents.len();
-        let generated = self
-            .get_boolean_attr(&filepath, "gengo-generated")?
-            .unwrap_or_else(|| self.is_generated(&filepath, contents));
-        let documentation = self
-            .get_boolean_attr(&filepath, "gengo-documentation")?
-            .unwrap_or_else(|| self.is_documentation(&filepath, contents));
-        let vendored = self
-            .get_boolean_attr(&filepath, "gengo-vendored")?
-            .unwrap_or_else(|| self.is_vendored(&filepath, contents));
+        let generated = attrs[1]
```

Comment:
Just a note since I found this bit a little hard to follow at first.
```suggestion
        // NOTE Unspecified attributes are filtered out, so if .is_set() is false, .is_unset() is true.
        let generated = attrs[1]
```

Reply:
This might be a hint that the API for how to deal with attribute values can be improved. I was also trying hard to assure it's actually doing the same as `git2` or, even better, `git` would do. I *think* it's correct, and if you have ideas on how to make it more intuitive, please let me know.